canton/community/ledger/ledger-api-core/src/main/scala/com/digitalasset/canton/platform/apiserver/services/admin/ApiPackageManagementService.scala

```diff
@@ -133,42 +137,87 @@ private[apiserver] final class ApiPackageManagementService private (
     } yield dar
   }
 
-  private def validateUpgrade(upgradingPackage: Dar[(Ref.PackageId, Ast.Package)])(implicit
+  private def lookupDar(pkgId: Ref.PackageId)(implicit
+                                                  loggingContext: LoggingContextWithTrace
+  ): Future[Option[(Ref.PackageId, Ast.Package)]] = {
+    for {
+      optArchive
+        <- packagesIndex.getLfArchive(pkgId)
+      optPackage
+        <- Future.fromTry {
+        optArchive.traverse(Decode.decodeArchive(_))
+          .handleError(Validation.handleLfArchiveError)
+      }
+    } yield optPackage
+  }
+
+  private def minimalVersionedDar(dar: (Ref.PackageId, Ast.Package))(implicit
+                                                                loggingContext: LoggingContextWithTrace
+  ): Future[Option[(Ref.PackageId, Ast.Package)]] = {
+    val (_, pkg) = dar
+    val pkgName = pkg.metadata.get.name
+    val pkgVersion = pkg.metadata.get.version
+    packageMetadataStore.getSnapshot.getUpgradablePackageMap.collect { case (pkgId, (`pkgName`, pkgVersion)) =>
+      (pkgId, pkgVersion)
+    }.minByOption { case (_, version) =>
+      pkgVersion < version
+    }.traverse { case (pId, _) => lookupDar(pId) }
+    .map(_.flatten)
+  }
+
+  private def maximalVersionedDar(dar: (Ref.PackageId, Ast.Package))(implicit
+                                                                loggingContext: LoggingContextWithTrace
+  ): Future[Option[(Ref.PackageId, Ast.Package)]] = {
+    val (_, pkg) = dar
+    val pkgName = pkg.metadata.get.name
+    val pkgVersion = pkg.metadata.get.version
+    packageMetadataStore.getSnapshot.getUpgradablePackageMap.collect { case (pkgId, (`pkgName`, pkgVersion)) =>
+      (pkgId, pkgVersion)
+    }.maxByOption { case (_, version) =>
+      pkgVersion > version
+    }.traverse { case (pId, _) => lookupDar(pId) }
+    .map(_.flatten)
+  }
+
+  private def typecheckUpgrades(optDar1: Option[(Ref.PackageId, Ast.Package)], optDar2: Option[(Ref.PackageId, Ast.Package)]): Future[Unit] = {
```

Comment:
Can we move this into Typecheck? I'd rather have as much of the logic in Typecheck as possible and only leave resolution of packages and converting of upgrade errors to the package service